torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+    def _write_back_current_shard(self):
+        for p in self.params:
+            if not p._is_sharded:
+                pass
+            chunks = p._full_param_padded.chunk(self.world_size)
+            if len(chunks) < (self.rank + 1):
+                # if there are not enough chunks, there's nothing to writeback here
+                pass
+        else:
+            chunk = chunks[self.rank]
+            p._local_shard.copy_(chunk)
+
+    @contextlib.contextmanager
+    def _summon_full_params(self, recurse: bool = True, volatile: bool = False) -> Generator:
+        """
+        A context manager to expose full params for the current FSDP instance.
+        Can be useful *after* forward/backward for a model to get the params for
+        additional processing or checking.
+        .. note:: This can be used on inner FSDPs.
+        .. note:: This can *not* be used within a forward or backward pass. Nor
+            can forward and backward be started from within this context.
+        .. note:: The full parameters will be freed after the context manager
+            exits; it is up to the caller to clone them if needed.
+        .. note:: The full parameters can be modified, but only the portion
+            corresponding to the local param shard will persist after the
+            context manager exits (unless ``volatile=True``, in which case
+            changes will be discarded).
+        Args:
+            recurse (bool, Optional): recursively summon all params for nested
+                FSDP instances (default: True)
+            volatile (bool, Optional): if ``True``, modifications to params are
+                discarded after the context manager exists;
+                enabling this can be slightly more efficient (default: False)
+        """
+        if recurse:
+            with contextlib.ExitStack() as stack:
+                # Summon all params for any nested FSDP instances.
+                for module in self.modules():
+                    if isinstance(module, FullyShardedDataParallel):
+                        stack.enter_context(module._summon_full_params(recurse=False, volatile=volatile))
+                # Yield to the caller, with full params in all nested instances.
+                yield
+            # Exiting from the ExitStack will re-shard params.
+            return
+        else:
+            torch.cuda.synchronize()
+            self._lazy_init()
+            self._assert_state([TrainingState_.IDLE])
+            # Set the state so that we assert when trying to go into
+            # forward/backward.
+            self.training_state = TrainingState_.SUMMON_FULL_PARAMS
+
+            self._rebuild_full_params()
+
+            try:
+                yield
+            finally:
+                if volatile is False:
+                    self._write_back_current_shard()
+                # FIXME should we respect #self.reshard_after_forward here too?
+                self._free_full_params()
```

Comment:
Is there a reason we can't just simply check the flag here?

Reply:
I was not sure on the expected behavior. It looks like checking it is the appropriate behavior.

Reply:
Looks like we should check for it. Fixed and added a test.